File: packages/python/trikhub/cli/commands/create_agent.py
"""trik create-agent — scaffold a minimal agent project."""

from __future__ import annotations

import re
import sys
from pathlib import Path

import click

from trikhub.cli.templates.agent_typescript import (
    CreateAgentConfig,
    generate_agent_typescript_project,
)
from trikhub.cli.templates.agent_python import generate_agent_python_project

NAME_PATTERN = re.compile(r"^[a-z][a-z0-9-]*$")

PROVIDERS = [
    ("openai", "OpenAI (gpt-4o-mini)"),
    ("anthropic", "Anthropic (claude-sonnet)"),
    ("google", "Google (gemini-2.0-flash)"),
]


def validate_name(name: str) -> str | None:
    if len(name) < 2 or len(name) > 50:
        return "Name must be 2-50 characters"
    if not NAME_PATTERN.match(name):
        return "Name must be lowercase, start with a letter, alphanumeric + dashes only"
    return None


@click.command("create-agent")
@click.argument("language", type=click.Choice(["ts", "typescript", "py", "python"]))
def create_agent_command(language: str) -> None:
    """Scaffold a minimal agent project ready to consume triks."""
    lang = "ts" if language in ("ts", "typescript") else "py"

    click.echo()
    click.echo(click.style("  Create a new Agent", bold=True))
    click.echo()

    # Project name
    name = click.prompt("Project name", default="my-agent").lower()
    error = validate_name(name)
    if error:
        click.echo(click.style(f"Error: {error}", fg="red"))
        sys.exit(1)

    # LLM provider
    click.echo("\nLLM Provider:")
    for i, (_, label) in enumerate(PROVIDERS, 1):
        click.echo(f"  {i}. {label}")
    provider_idx = click.prompt("Provider (number)", type=int, default=1)
    provider = PROVIDERS[max(0, min(provider_idx - 1, len(PROVIDERS) - 1))][0]

    # Path
    target_dir = Path.cwd() / name
    use_current = click.confirm(f"Create in ./{name}?", default=True)
    if not use_current:
        custom = click.prompt("Enter path", default=f"./{name}")
        target_dir = Path.cwd() / custom

    if target_dir.exists():
        click.echo(click.style(f"\nDirectory already exists: {target_dir}", fg="red"))
        sys.exit(1)

    click.echo()

    # Generate files
    config = CreateAgentConfig(name=name, provider=provider)

    if lang == "ts":
        files = generate_agent_typescript_project(config)
    else:
        files = generate_agent_python_project(config)

    # Write files
    for f in files:
        file_path = target_dir / f.path
        file_path.parent.mkdir(parents=True, exist_ok=True)
        file_path.write_text(f.content, encoding="utf-8")

    click.echo(click.style("  Your agent is ready!", fg="green", bold=True))
    click.echo()
    click.echo(click.style("  Next steps:", dim=True))
    click.echo(f"    cd {name}")
    click.echo("    cp .env.example .env")
    click.echo("    # Add your API key to .env")

    if lang == "py":
        click.echo("    python -m venv .venv && source .venv/bin/activate")
        click.echo("    pip install -e .")
        click.echo("    python cli.py")
    else:
        click.echo("    npm install && npm run dev")

    click.echo()
    click.echo(click.style("  Install triks:", dim=True))
    click.echo("    trik install @scope/trik-name")
    click.echo()

File: packages/js/cli/src/commands/create-agent.ts
/**
 * trik create-agent command
 *
 * Scaffolds a minimal agent project ready to consume triks.
 */

import { existsSync } from 'node:fs';
import { mkdir, writeFile } from 'node:fs/promises';
import { join, resolve } from 'node:path';
import chalk from 'chalk';
import ora from 'ora';
import { input, select } from '@inquirer/prompts';
import { generateAgentTypescriptProject } from '../templates/agent-typescript.js';
import { generateAgentPythonProject } from '../templates/agent-python.js';
import type { CreateAgentConfig } from '../templates/agent-typescript.js';

type Language = 'ts' | 'py';

/**
 * Validate project name — same rules as trik init.
 */
function validateProjectName(name: string): string | true {
  if (name.length < 2 || name.length > 50) {
    return 'Name must be 2-50 characters';
  }
  if (!/^[a-z]/.test(name)) {
    return 'Name must start with a letter';
  }
  if (!/^[a-z][a-z0-9-]*$/.test(name)) {
    return 'Name must be lowercase, alphanumeric + dashes only';
  }
  return true;
}

/**
 * Parse and normalize language argument.
 */
function parseLanguage(lang: string): Language | null {
  const normalized = lang.toLowerCase();
  if (normalized === 'ts' || normalized === 'typescript') {
    return 'ts';
  }
  if (normalized === 'py' || normalized === 'python') {
    return 'py';
  }
  return null;
}

export async function createAgentCommand(languageArg: string): Promise<void> {
  const spinner = ora();

  try {
    // Validate language
    const language = parseLanguage(languageArg);
    if (!language) {
      console.log(chalk.red(`Invalid language: ${languageArg}`));
      console.log(chalk.dim('Supported languages: ts (TypeScript) or py (Python)'));
      process.exit(1);
    }

    console.log();
    console.log(chalk.bold('  Create a new Agent'));
    console.log();

    // Interactive prompts
    const name = await input({
      message: 'Project name:',
      default: 'my-agent',
      validate: validateProjectName,
      transformer: (value) => value.toLowerCase(),
    });

    const provider = await select<'openai' | 'anthropic' | 'google'>({
      message: 'LLM Provider:',
      choices: [
        { value: 'openai', name: 'OpenAI (gpt-4o-mini)' },
        { value: 'anthropic', name: 'Anthropic (claude-sonnet)' },
        { value: 'google', name: 'Google (gemini-2.0-flash)' },
      ],
      default: 'openai',
    });

    // Path selection
    const pathChoice = await select<'current' | 'other'>({
      message: 'Where to create the project?',
      choices: [
        { value: 'current', name: `Current folder (./${name})` },
        { value: 'other', name: 'Other location...' },
      ],
    });

    let targetDir: string;
    if (pathChoice === 'current') {
      targetDir = resolve(process.cwd(), name);
    } else {
      const customPath = await input({
        message: 'Enter path:',
        default: `./${name}`,
      });
      targetDir = resolve(process.cwd(), customPath);
    }

    // Check if directory exists
    if (existsSync(targetDir)) {
      console.log();
      console.log(chalk.red(`Directory already exists: ${targetDir}`));
      process.exit(1);
    }

    console.log();
    spinner.start('Creating agent project...');

    // Build config
    const config: CreateAgentConfig = { name, provider };

    // Generate project files
    const files = language === 'ts'
      ? generateAgentTypescriptProject(config)
      : generateAgentPythonProject(config);

    // Create target directory and write all files
    await mkdir(targetDir, { recursive: true });

    for (const [relativePath, content] of Object.entries(files)) {
      const filePath = join(targetDir, relativePath);
      const fileDir = join(filePath, '..');
      await mkdir(fileDir, { recursive: true });
      await writeFile(filePath, content, 'utf-8');
    }

    spinner.succeed('Agent project created');

    // Install dependencies (TypeScript only)
    let packageManager = 'npm';
    if (language === 'ts') {
      const { execSync } = await import('node:child_process');
      try {
        execSync('pnpm --version', { stdio: 'ignore' });
        packageManager = 'pnpm';
      } catch {
        // pnpm not available, use npm
      }

      spinner.start(`Installing dependencies with ${packageManager}...`);
      try {
        execSync(`${packageManager} install`, { cwd: targetDir, stdio: 'ignore' });
        spinner.succeed('Dependencies installed');
      } catch {
        spinner.warn(`Failed to install dependencies. Run \`${packageManager} install\` manually.`);
      }
    }

    // Show success message
    console.log();
    console.log(chalk.green.bold('  Your agent is ready!'));
    console.log();
    console.log(chalk.dim('  Next steps:'));
    console.log(`    cd ${name}`);
    console.log('    cp .env.example .env');
    console.log('    # Add your API key to .env');

    if (language === 'py') {
      console.log('    python -m venv .venv && source .venv/bin/activate');
      console.log('    pip install -e .');
      console.log('    python cli.py');
    } else {
      console.log(`    ${packageManager === 'pnpm' ? 'pnpm' : 'npm run'} dev`);
    }

    console.log();
    console.log(chalk.dim('  Install triks:'));
    console.log('    trik install @scope/trik-name');
    console.log();
  } catch (error) {
    spinner.fail('Failed to create agent project');
    if (error instanceof Error) {
      // Handle user cancellation gracefully
      if (error.message.includes('User force closed')) {
        process.exit(0);
      }
      console.error(chalk.red(error.message));
    }
    process.exit(1);
  }
}
